compiler/qsc_ast/src/ast.rs

```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
         }
     }
 }
+
+/// A [`TypeParameter`] is a generic type variable with optional bounds (constraints).
+#[derive(Default, PartialEq, Eq, Clone, Hash)]
+pub struct TypeParameter {
+    /// Class constraints specified for this type parameter -- any type variable passed in
+    /// as an argument to these parameters must satisfy these constraints.
+    pub constraints: ClassConstraints,
+    /// The name of the type parameter.
+    pub ty: Ident,
+    /// The span of the full type parameter, including its name and its constraints.
+    pub span: Span,
+}
+
+impl WithSpan for TypeParameter {
+    fn with_span(self, span: Span) -> Self {
+        Self {
+            span,
+            ty: self.ty.with_span(span),
+            ..self
+        }
+    }
+}
+
+impl TypeParameter {
+    /// Instantiates a new `TypeParameter` with the given type name, constraints, and span.
+    #[must_use]
+    pub fn new(ty: Ident, bounds: ClassConstraints, span: Span) -> Self {
+        Self {
+            ty,
+            constraints: bounds,
+            span,
+        }
+    }
+}
+
+impl std::fmt::Debug for TypeParameter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // 'A: Eq + Ord + Clone
+        write!(
+            f,
+            "{}{}{:?}",
+            self.ty,
+            if self.constraints.0.is_empty() {
+                ""
+            } else {
+                ": "
+            },
+            self.constraints
+        )
+    }
+}
```

Comment:
This and the `Debug` impl for `ClassConstraint` are the first custom `Debug` impls in this file. The norm is to define `Display` for AST nodes, and use that in tests / AST dumps . We can debate whether that's the "right" way but it's a pretty entrenched pattern. Can you just make these `Display` impls maybe?